[RFC] Simplify UI mode change API and use underline cursor in Replace mode

osandov opened the pull request:
Vim as of recently supports having a separate cursor shape for Replace mode (e.g., an underline, like in gVim). I contributed the original patch to Vim; now that I want to try out Neovim, I missed this feature, so I figured I should contribute it. It's not a mechanical forward-port, as the TUI code is very different in Neovim.

Problem:    There is no way to use a different in Replace mode for a terminal.
Solution:   Add t_SR. (Omar Sandoval)

https://github.com/vim/vim/commit/v7-4-687

The TUI code has been heavily refactored (see esp. 25ceadab37edb ("tui:
Remove support for overriding escape sequences with nvim options")), so
this forward-port required some translation.

justinmk commented:
Thanks a lot! Doesn't need to be marked as a vim-patch in the commit message, since you are the copyright owner, it is licensed under Apache 2.0. However, `version.c` could still be marked as satisfied or possibly NA.

I guess we can close this: https://github.com/neovim/neovim/pull/2066 ? Maybe you can comment on the differences there?

justinmk commented:
LGTM

justinmk commented:
I will mention though, the change @tarruda mentioned here:

https://github.com/neovim/neovim/pull/2066#issuecomment-76369286

is still an avenue we want to explore if you feel like it.

osandov committed:
Simplify UI API on mode change

Currently, there are two functions in the UI API that are called when
the mode changes: insert_mode() and normal_mode(). These can be folded
into a single change_mode() entrypoint which can do whatever it wants
based on the mode it is passed, limited to INSERT and NORMAL for now.

osandov committed:
Use underline cursor in TUI in Replace mode

This is a port of my original contribution to Vim, added in 7.4.687
(https://github.com/vim/vim/commit/v7-4-687). The TUI code has been
heavily refactored (see esp. 25ceadab37edba13f5afa78d8b4723da03ef35f0),
so this required some translation, but the logic is the same.

osandov commented:
Alright, here's a go at that.

fwalch commented:
Can you update the PR title?

fmoralesc commented:
Once this gets included, we could look into reimplementing the `guicursor` feature for the TUI (someone mentioned it in the irc room).

tarruda commented on src/nvim/msgpack_rpc/remote_ui.c:
```diff
@@ -214,16 +213,11 @@ static void remote_ui_mouse_off(UI *ui)
-{
-  Array args = ARRAY_DICT_INIT;
-  push_call(ui, "normal_mode", args);
+  ADD(args, INTEGER_OBJ(mode));
```
I'd prefer if we sent a "normal"/"insert"/"replace" string instead of an integer in this case. For builtin UIs(tui.c) it is ok to use an integer since it has direct access to neovim headers, but remote uis can be written in any language, so it's simpler to use string identifiers.

tarruda commented:
Other than the suggested change to remote_ui.c, LGTM

osandov commented:
Fixed and rebased on current master. Thanks!

justinmk commented:
Merged, thanks!

I renamed "change_mode" to "mode_change". Although we have not been consistent about it, we should follow a "noun verb" pattern since nouns tend to be more distinctive. (Events should also not be commands, but that's another discussion)

justinmk commented:
@carlosdcastillo @coolwanglu @qvacua @rogual @/equalsraf @oakes @splinterofchaos @Pireax @jebberjeb @awetzel @fmoralesc @myitcv @saep @fdinoff @bfredl @fritzy @ngroup @jalvesaq @alexgenco 

This change may break your client. It removes the `insert_mode` and `normal_mode` UI messages, and added a `mode_change` message. We would appreciate feedback on which clients were affected, and how difficult it was to fix. (If you weren't affected, feel free to ignore me, and we will count that as "not affected")

bfredl commented:
Maybe ui clients should be advised to ignore ui messages they don't recognize (rather than error out). Perhaps a bad example with a renamed event, but that would allow us to add new events with new information without breakage.

coolwanglu commented:
@justinmk Thanks for pinging me, this is cool! I wonder if there's a group tag or mailing list to simplify this, e.g. @gui_imp

Also removing support for old messages would break if the user is not using the latest commit, but it might be OK since there's no stable release yet?

equalsraf commented:
@justinmk mostly good on my front, the cursor stopped changing with mode (which is to be expected). In neovim-qt unknown messages are just dropped anyway and logged in debug builds.

There was (I think) an old discussion about putting some Neovim version info in the API metadata - that should help verifying compatibility.

rogual commented:
@justinmk thanks for the heads-up. This was an easy change to accommodate in Neovim.app (https://github.com/rogual/neovim-dot-app/commit/e674855473fae3aa1686e77c1f8161dc44fbe5db if you're interested). I'm also ignoring unknown messages so no major breakage even without that patch.

@coolwanglu I tend to agree re. backcompat pre 1.0, and didn't add any to my client when making this change.
